src/D3MHub.sol

```diff
@@ -0,0 +1,603 @@
+// SPDX-FileCopyrightText: © 2021-2022 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021-2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import "./pools/ID3MPool.sol";
+import "./plans/ID3MPlan.sol";
+
+interface VatLike {
+    function debt() external view returns (uint256);
+    function hope(address) external;
+    function nope(address) external;
+    function ilks(bytes32) external view returns (uint256, uint256, uint256, uint256, uint256);
+    function Line() external view returns (uint256);
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function gem(bytes32, address) external view returns (uint256);
+    function live() external view returns (uint256);
+    function slip(bytes32, address, int256) external;
+    function move(address, address, uint256) external;
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function grab(bytes32, address, address, address, int256, int256) external;
+    function fork(bytes32, address, address, int256, int256) external;
+    function suck(address, address, uint256) external;
+}
+
+interface EndLike {
+    function debt() external view returns (uint256);
+    function skim(bytes32, address) external;
+}
+
+interface DaiJoinLike {
+    function dai() external view returns (address);
+    function join(address, uint256) external;
+    function exit(address, uint256) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external returns (bool);
+}
+
+/**
+    @title D3M Hub
+    @notice This is the main D3M contract and is responsible for winding and
+    unwinding pools, interacting with DSS and tracking the plans and pools and
+    their states.
+*/
+contract D3MHub {
+
+    // --- Auth ---
+    /**
+        @notice Maps address that have permission in the Pool.
+        @dev 1 = allowed, 0 = no permission
+        @return authorization 1 or 0
+    */
+    mapping (address => uint256) public wards;
+
+    address public vow;
+    EndLike public end;
+    uint256 public locked;
```

Comment:
`locked` has usually `internal` visibility

Reply:
I think we wanted to keep it public for if there's ever a situation it needs to be checked by an integrating contract to prevent some reentrancy situation we are currently not awawre of.

Reply:
Yeah I think this was discussed in one of the calls. I don't have a strong preference to be honest, so if anyone thinks is better to have it public it doesn't really harm.